src/undo/EraseUndoAction.cpp

```diff
@@ -1,86 +1,48 @@
-        }
-        delete e;
-    }
-    g_list_free(this->edited);
-    this->edited = nullptr;
-}
-
-void EraseUndoAction::addOriginal(Layer* layer, Stroke* element, int pos) {
-    this->original = g_list_insert_sorted(this->original, new PageLayerPosEntry<Stroke>(layer, element, pos),
-                                          reinterpret_cast<GCompareFunc>(PageLayerPosEntry<Stroke>::cmp));
-}
+void EraseUndoAction::addOriginal(Layer* layer, Stroke* element, int pos) { original.emplace(layer, element, pos); }
 
-void EraseUndoAction::addEdited(Layer* layer, Stroke* element, int pos) {
-    this->edited = g_list_insert_sorted(this->edited, new PageLayerPosEntry<Stroke>(layer, element, pos),
-                                        reinterpret_cast<GCompareFunc>(PageLayerPosEntry<Stroke>::cmp));
-}
+void EraseUndoAction::addEdited(Layer* layer, Stroke* element, int pos) { edited.emplace(layer, element, pos); }
 
 void EraseUndoAction::removeEdited(Stroke* element) {
-    for (GList* l = this->edited; l != nullptr; l = l->next) {
-        auto* p = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        if (p->element == element) {
-            this->edited = g_list_delete_link(this->edited, l);
-            delete p;
-            p = nullptr;
+    for (auto entryIter = edited.begin(); entryIter != edited.end(); entryIter++) {
+        if (entryIter->element == element) {
+            edited.erase(entryIter);
             return;
         }
     }
 }
 
 void EraseUndoAction::finalize() {
-    for (GList* l = this->original; l != nullptr;) {
-        auto* p = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        GList* del = l;
-        l = l->next;
-
-        if (p->element->getPointCount() == 0) {
-            this->edited = g_list_delete_link(this->edited, del);
-            delete p;
-            p = nullptr;
+    for (auto entryIter = original.begin(); entryIter != original.end(); entryIter++) {
+        if (entryIter->element->getPointCount() == 0) {
+            // TODO (Marmare314): is this really expected behaviour?
+            continue;
         } else {
-
             // Remove the original and add the copy
-            int pos = p->layer->removeElement(p->element, false);
-
-            EraseableStroke* e = p->element->getEraseable();
-            GList* stroke = e->getStroke(p->element);
-            for (GList* ls = stroke; ls != nullptr; ls = ls->next) {
-                auto* copy = static_cast<Stroke*>(ls->data);
-                p->layer->insertElement(copy, pos);
-                this->addEdited(p->layer, copy, pos);
+            int pos = static_cast<int>(entryIter->layer->removeElement(entryIter->element, false));
+
+            EraseableStroke* e = entryIter->element->getEraseable();
+            std::vector<std::unique_ptr<Stroke>> strokeList = e->getStroke(entryIter->element);
+            for (auto& stroke: strokeList) {
+                // TODO (Marmare314): should use unique_ptr in layer
```

Comment:
Thank you for the todo :); Yes that is a refactoring, which should go into another pr :)